### PR TITLE
Restrict user profile write acces to owner and super user

### PR DESCRIPTION
**Pull request by rszwajko**
Before, every admin user has full write access to user profiles of other users. 
After this patch only super users will be able to modify user profiles of other users.
Admin users will have read only access.

### CODE REVIEW

**Review by sgratch — COMMENTED** [review]
LGTM
The only missing part is handling the UI side for the new account settings tab by displaying a popup error message in case of a non superuser trying to delete for someone else, e.g.

image

**Comment by rszwajko** [comment]
@sgratch 
Similar as other errors on this screen this is handled by the notifications system (see below)
image

**Comment by rszwajko** [comment]
The functionality can be tested without PR #130 via REST API
```
curl -v \
--request DELETE \
--header 'Version: 4' \
--header 'Accept: application/xml' --header 'Content-Type: application/xml' \
--user "adminNotSuperUser@internal:password" \
{server}/ovirt-engine/api/users/{other_user_id}/options/{some_option_id}
```
Example result:
```
<fault>
    <detail>[User is not authorized to perform this action.]</detail>
    <reason>Operation Failed</reason>
</fault>

```

**Comment by rszwajko** [comment]
@sgratch 
> other such errors of "User is not authorized to perform this action" generated by the Administration->Users dialog are displayed using a popup error message. Why not running the same for this new tab?
>

This comment applies to error handling in  PR #130. In March we discussed [this topic](https://github.com/oVirt/ovirt-engine/pull/130#issuecomment-1079928575) and apart of wording there were no issues raised. 
In Web Admin we have either default (generic) error handlers or custom handlers. On the screen added in PR #130  we are using a custom error handler.

**Review by sgratch — COMMENTED** [review]
@rszwajko comments regarding the tests, except that LGTM

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/RemoveUserProfilePropertyCommandTest.java [hunk 1]
```diff
@@ -28,6 +31,9 @@ class RemoveUserProfilePropertyCommandTest extends BaseCommandTest {
     @Mock
     private DbUserDao userDaoMock;
 
+    @Mock
+    private Predicate<DbUser> isSystemSuperUserPredicate;
```
Comment (sgratch) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/RemoveUserProfilePropertyCommandTest.java:35]: The test condition `when(isSystemSuperUserPredicate.test(...).thenReturn(true)` seems missing for `isSystemSuperUserPredicate`

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/UpdateUserProfilePropertyCommandTest.java [hunk 2]
```diff
@@ -32,6 +34,9 @@ class UpdateUserProfilePropertyCommandTest extends BaseCommandTest {
     @Mock
     private DbUserDao userDaoMock;
 
+    @Mock
+    private Predicate<DbUser> isSystemSuperUserPredicate;
```
Comment (sgratch) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/UpdateUserProfilePropertyCommandTest.java:38]: The test condition `when(isSystemSuperUserPredicate.test(....).thenReturn(true)` seems missing for `isSystemSuperUserPredicate`

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/RemoveUserProfilePropertyCommandTest.java (hunk 1 above)
Comment (rszwajko) [reply backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/RemoveUserProfilePropertyCommandTest.java:35]: `isSystemSuperUserPredicate.test()` returns correctly `false` as we have here no super user scenario. This scenario is tested in `AddUserProfilePropertyCommandTest.addSinglSshPropToEmptyProfileForAnotherUser`.

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/UpdateUserProfilePropertyCommandTest.java (hunk 2 above)
Comment (rszwajko) [reply backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/UpdateUserProfilePropertyCommandTest.java:38]: As above: we have no super user scenario here.

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/validator/UserProfileValidator.java [hunk 3]
```diff
@@ -40,14 +40,19 @@ public ValidationResult propertyProvided(UserProfileProperty property) {
                 .when(property == null);
     }
 
-    public ValidationResult authorized(DbUser currentUser, Guid targetUserId) {
+    public ValidationResult authorizedToModify(DbUser currentUser, Guid targetUserId, boolean isSuperUser) {
```
Comment (ljelinkova) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/validator/UserProfileValidator.java:43]: `isSuperUser` is not really correct as you call this method below from the method `authorized` where the user is only `admin`. What about something like `sufficientAdminRights`?

Also, I would keep the method's name to `authorized` as it is quite general and is not used just for modification.

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/validator/UserProfileValidator.java [hunk 4]
```diff
@@ -40,14 +40,19 @@ public ValidationResult propertyProvided(UserProfileProperty property) {
...
         return ValidationResult
                 .failWith(EngineMessage.USER_NOT_AUTHORIZED_TO_PERFORM_ACTION)
                 .when(currentUser == null
                         || currentUser.getId() == null
-                        || !Objects.equals(currentUser.getId(), targetUserId) && !currentUser.isAdmin());
+                        || !Objects.equals(currentUser.getId(), targetUserId) && !isSuperUser);
+    }
+
+    public ValidationResult authorized(DbUser currentUser, Guid targetUserId) {
```
Comment (ljelinkova) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/validator/UserProfileValidator.java:51]: I would remove this method and call the general `authorized` method where needed with `currentUser != null && currentUser.isAdmin()` argument.

**Review by ljelinkova — APPROVED** [review]

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/RemoveUserProfilePropertyCommandTest.java (hunk 1 above)
Comment (sgratch) [reply backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/RemoveUserProfilePropertyCommandTest.java:35]: @rszwajko so why declaring `isSystemSuperUserPredicate` at all if `isSystemSuperUserPredicate.test()` is not used for `RemoveUserProfilePropertyCommandTest` (and same for `UpdateUserProfilePropertyCommandTest`)?

Comment (rszwajko) [reply backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/RemoveUserProfilePropertyCommandTest.java:35]: @sgratch 
It is used and returns `false` as we are testing a standard user scenario there.

**Review by sgratch — APPROVED** [review]

**Pull request merged**